examples/improved_wgan.py

```diff
@@ -0,0 +1,263 @@
+    parser.add_argument("--output_dir", "-o", required=True, help="Directory to output generated files to")
+    args = parser.parse_args()
+
+    # First we load the image data, reshape it and normalize it to the range [-1, 1]
+    (X_train, y_train), (X_test, y_test) = mnist.load_data()
+    X_train = np.concatenate((X_train, X_test), axis=0)
+    X_train = X_train.reshape((X_train.shape[0], X_train.shape[1], X_train.shape[2], 1))
+    X_train = (X_train.astype(np.float32) - 127.5) / 127.5
+
+    # Now we initialize the generator and discriminator.
+    generator = make_generator()
+    discriminator = make_discriminator()
+
+    # The generator_model is used when we want to train the generator layers.
+    # As such, we ensure that the discriminator layers are not trainable.
+    # Note that once we compile this model, updating .trainable will have no effect within it. As such, it
+    # won't cause problems if we later set discriminator.trainable = True for the discriminator_model, as long
+    # as we compile the generator_model first.
+    for layer in discriminator.layers:
+        layer.trainable = False
+    discriminator.trainable = False
+    generator_input = Input(shape=(100,))
+    generator_layers = generator(generator_input)
+    discriminator_layers_for_generator = discriminator(generator_layers)
+    generator_model = Model(inputs=[generator_input], outputs=[discriminator_layers_for_generator])
+    # We use the Adam paramaters from Gulrajani et al.
+    generator_model.compile(optimizer=Adam(0.0001, beta_1=0.5, beta_2=0.9), loss=wasserstein_loss)
+
+    # Now that the generator_model is compiled, we can make the discriminator layers trainable.
+    for layer in discriminator.layers:
+        layer.trainable = True
+    for layer in generator.layers:
+        layer.trainable = False
+    discriminator.trainable = True
+    generator.trainable = False
+
+    # The discriminator_model is more complex. It takes both real image samples and random noise seeds as input.
+    # The noise seed is run through the generator model to get generated images. Both real and generated images
+    # are then run through the discriminator. Although we could concatenate the real and generated images into a
+    # single tensor, we don't (see model compilation for why).
+    real_samples = Input(shape=X_train.shape[1:])
+    generator_input_for_discriminator = Input(shape=(100,))
+    generated_samples_for_discriminator = generator(generator_input_for_discriminator)
+    discriminator_output_from_generator = discriminator(generated_samples_for_discriminator)
+    discriminator_output_from_real_samples = discriminator(real_samples)
+
+    # We also need to generate weighted-averages of real and generated samples, to use for the gradient norm penalty.
+    averaged_samples = RandomWeightedAverage()([real_samples, generated_samples_for_discriminator])

```

Comment:
This is definitely the most confusing part of the code, but I'm fairly sure I've implemented it correctly. You are correct that I'm averaging real and generated samples, though.

The core idea is that you want the gradient norm to be close to 1 everywhere; this is the essence of the Lipschitz constraint that must be met for WGANs to train properly. However, it is impossible to guarantee that the gradient norm for every possible input will be close to 1. The compromise used in the paper is that you "Sample along straight lines" between real and generated samples. Equations (4) and (5) on page 6 describe how they do this, which you will see is equivalent to the randomly-weighted average between real and generated samples from my function. It is effectively equivalent to plotting a line between a real sample and a generated sample, and sampling a random point on that line with uniform probability.

The reason the randomly-averaged functions are passed through the discriminator afterwards is not because we particularly care about their output. Instead, we care about the _gradient_ we get, because we want to ensure its norm stays close to 1. When the model is compiled, you'll notice the loss for the randomly-averaged samples is not the wasserstein_loss, but the partial_gp_loss function.